flang/test/Lower/dummy-procedure-character.f90

```diff
@@ -0,0 +1,172 @@
+! Test lowering of character function dummy procedure. The length must be
+! passed along the function address.
+! RUN: bbc -emit-fir %s -o - | FileCheck %s
+
+! -----------------------------------------------------------------------------
+!     Test passing a character function as dummy procedure
+! -----------------------------------------------------------------------------
+
+! CHECK-LABEL: func @_QPcst_len
+subroutine cst_len()
+  interface
+    character(7) function bar1()
+    end function
+  end interface
+! CHECK:  %[[VAL_0:.*]] = fir.address_of(@_QPbar1) : (!fir.ref<!fir.char<1,?>>, index) -> !fir.boxchar<1>
+! CHECK:  %[[VAL_1:.*]] = arith.constant 7 : i64
+! CHECK:  %[[VAL_2:.*]] = fir.convert %[[VAL_0]] : ((!fir.ref<!fir.char<1,?>>, index) -> !fir.boxchar<1>) -> !fir.ref<!fir.char<1,?>>
```

Comment:
This cast is lossy and odd looking to me. It is lossy in that we have a function pointer and are converting it to a data pointer, which loses information that the pointer is to code and not data. Subsequently, the IR has no way to reason about this value correctly.

Reply:
Makes sense, what do you see as an alternative here ?  It is currently not possible to create a fir.embox_char with a function address. Should we allow that to avoid the cast or should we create a new operation here (other alternatives welcomed if you see any) ?